type-auto-algorithm.md

```diff
@@ -0,0 +1,55 @@
+# `--type=auto` Module Type Detection Algorithm
+
+This is the algorithm, in psuedocode, for the `--type=auto`/`-a` Node startup option described [here](./README.md). It is based on detailed syntax rules outlined [here](https://github.com/nodejs/node-eps/issues/57#issuecomment-300870976).
+
+This algorithm relies on the fact that it is evaluating an initial entry point—the very first file loaded for a program. Affirmatively detecting CommonJS would be virtually impossible otherwise, as it would no longer be a safe assumption that a reference to a globally defined `require` implies that a file is CommonJS (since a file loaded earlier in the program could have defined `require` in the global scope). Additionally, many files (either CommonJS or ESM) that are not entry points are often ambiguous, for example containing only `console.log` statements or references to user-defined globals. For this algorithm to be both confident (no false positives for ESM, no realistic false positives for CommonJS) and inclusive (few files judged ambiguous) it must be restricted to initial entry points.
+
+## Step 1: Parse
+
+The initial entry point file must be parsed so that its syntax may be analyzed.
+
+## Step 2: Look for signifiers of ESM JavaScript
+
+If the source code contains either of the following:
+
+- An `import` statement, like `import "x";`
+
+- An `export` statement, like `export {};`
+
+Then the file is unambiguously ESM. Stop and execute it as ESM JavaScript.
+
+These are the only two syntactical elements that are unambiguously ESM. The `import` expression, `import("x")`, is allowed in CommonJS.
+
+If the source code did not contain an ESM signifier, continue:
+
+## Step 3: Look for signifiers of CommonJS JavaScript
+
+If the source code contains any of the following:
+
+- A reference to a global variable named `__filename`, without code defining the variable.
```

Comment:
```suggestion
- A static reference, unguarded by `typeof`, to a global variable named `__filename`, without code defining the variable.
```
and the same change for the others.

also, maybe a reference to `eval` could just bail out of the auto detection at all, since all bets are off then?